packages/api-openfin/src/main-process.ts

```diff
@@ -0,0 +1,144 @@
+import MessageService from './message-service';
+
+// Code that is "evaled" when the main window has been opened, sets up
+// all the InterApplicationBus listeners for window events to keep track of state
+const mainWindowCode = () => {
+  const childTree = [];
+
+  const addParent = (parentName, windowName, tree) => {
+    let win = tree.find(w => w.name === parentName);
+    if (win) {
+      win.children.push({
+        name: windowName,
+        children: []
+      });
+      return true;
+    } else {
+      return tree.some((childWin) => {
+        return addParent(parentName, windowName, childWin.children);
+      });
+    }
+  };
+
+  const getParentWindow = (name, parentName, tree) => {
+    let win = tree.find(w => w.name === name);
+    if (win) {
+      return parentName;
+    } else {
+      let parentName = null;
+      tree.some((childWin) => {
+        parentName = getParentWindow(name, childWin.name, childWin.children);
+        return parentName !== null;
+      });
+      return parentName;
+    }
+  };
+
+  const getChildWindows = (name, tree) => {
+    let win = tree.find(w => w.name === name);
+    if (win) {
+      return win.children.map(c => c.name);
+    } else {
+      let childNames = [];
+      tree.some((childWin) => {
+        childNames = getChildWindows(name, childWin.children);
+        return childNames && childNames.length > 0;
+      });
+      return childNames;
+    }
+  };
+
+  const deleteWindow = (name, tree) => {
+    let index = tree.findIndex(w => w.name === name);
+    if (index >= 0) {
+      tree.splice(index, 1);
+    } else {
+      tree.forEach((childWin) => {
+        deleteWindow(name, childWin.children);
+      });
+    }
+  };
+
+  const closeChildren = (uuid) => {
+    const childUuids = getChildWindows(uuid, childTree);
+    childUuids.forEach((child) => {
+      fin.desktop.Application.wrap(child).close(true);
+      closeChildren(child);
+    });
+  }
+
+  fin.desktop.InterApplicationBus.subscribe('*', 'ssf-new-window', (data) => {
```

Comment:
The problem doing this is that multiple windows need to be able to send to this window, and because each window is now a new application they all have a different UUID. Especially as before this point, we don't know a new window exists (as this is telling the `mainWindow` of a new window) so having a list of UUIDs in our application wouldn't help. Unless you can think of a different way around it?